Ensures that mounted routers have a leading slash

atheriel opened the pull request:
This should fix #476. Since Plumber already ensures that there is a trailing slash, this just does the same for a leading slash as well, rather than issue an error (which was the other suggested solution).

The examples from that thread now work correctly.

PR task list:
- [x] Update NEWS
- [x] Add tests
- [x] Update documentation with `devtools::document()`

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/rstudio/plumber?pullRequest=501) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/rstudio/plumber?pullRequest=501) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/rstudio/plumber?pullRequest=501) it.</sub>

wch commented on R/plumber.R:
```diff
@@ -322,7 +322,8 @@ plumber <- R6Class(
     mount = function(path, router){
-      path <- sub("([^/])$", "\\1/", path)
+      # Ensure that the path has both a leading and trailing slash.
+      path <- sub("^/?(.*)([^/])$", "/\\1\\2/", path)
```
This won't catch the case where `path` has a trailing slash but no leading slash.

```R
sub("^/?(.*)([^/])$", "/\\1\\2/", "abc/")
#> [1] "abc/"
```

In general, I get a little nervous when there's conditional logic in a regexp, so I would try to avoid it if possible, especially if there are multiple conditions you're testing for. I'd suggest using two separate `if` statements, with something like this: 

```R
if (substr(path, 1, 1) != "/") {
  path <- paste0("/", path)
}
if (substr(path, nchar(path), nchar(path)) != "/") {
  path <- paste0(path, "/")
}
```

It's very verbose, but at least it's clear what it's doing.

atheriel replied:
I prefer `substr` or even `startsWith()` as well, I just stuck to a regex to minimize changes.

> This won't catch the case where path has a trailing slash but no leading slash.

Ah. I thought I had tested that, but clearly not.

schloerke replied:
+1 to `startsWith`

`?base::startsWith`
```
‘startsWith()’ is equivalent to but much faster than

     
     substring(x, 1, nchar(prefix)) == prefix
```

There's also an `endsWith`. TIL

wch commented on tests/testthat/test-plumber.R:
```diff
@@ -150,9 +150,11 @@ test_that("mounts can be read correctly", {
 
   stat <- PlumberStatic$new(".")
   pr$mount("/static", stat)
+  pr$mount("missing-slash", stat)
```
It would be good to have two more cases, with trailing slashes.

atheriel replied:
Will do.

atheriel commented:
Should be good to go now. I added some additional tests and moved to using `startsWith()` and `endsWith()` instead of a regex. I've listed @wch as a co-author on the commit, since he basically composed the new solution.
